### PR TITLE
MIG-119 - Introduce migration pipeline

### PR DESCRIPTION
**Pull request by pranavgaikwad**
This PR is part of the progress reporting improvements in mig-controller. It introduces a new field in `Phase` struct to assign a `Phase` with a logically related high level _Step_:

```go
// Phase
type Phase struct {
	[...]
        Step string
}
```

High level steps introduced:

```
Prepare
Backup
StageBackup
StageRestore
Restore
Final
```

## Pipeline

This PR also introduces a new field in _MigMigrationStatus_ struct to capture the progress of an ongoing migration through different _Steps_.

## Phase descriptions

Phase description work is cherry-picked from this PR: https://github.com/konveyor/mig-controller/pull/656

Originally, the proposed field for description was `Status.Description`. This PR instead uses the `message` field of current step in the pipeline. 

### Pipeline in action

Example 1: Failed Migration

```yml
status:
  conditions:
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:37:01Z"
    message: '[1] Stage pods created.'
    status: "True"
    type: StagePodsCreated
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:37:32Z"
    message: 'The migration has failed.  See: Errors.'
    reason: StageBackup
    status: "True"
    type: Failed
  errors:
  - 'PVC: mssql-persistent/mssql-pvc, deleted.'
  itinerary: Failed
  observedDigest: 9834d071975562d5e2c2eb855bca6950711ded8a0e45af5307fa56cd0f5ba3c7
  phase: Completed
  pipeline:
  - completed: "2020-10-29T21:35:53Z"
    message: Completed
    name: Prepare
    started: "2020-10-29T21:33:02Z"
  - completed: "2020-10-29T21:36:58Z"
    message: Completed
    name: Backup
    started: "2020-10-29T21:35:53Z"
  - completed: "2020-10-29T21:37:32Z"
    failed: true
    message: Failed
    name: StageBackup
    started: "2020-10-29T21:36:58Z"
  - message: Not started
    name: StageRestore
  - message: Not started
    name: Restore
  - completed: "2020-10-29T21:37:34Z"
    message: Completed
    name: Final
    started: "2020-10-29T21:37:32Z"
  startTimestamp: "2020-10-29T21:33:02Z"
```

Example 2 : Successful migration

```yml
status:
  conditions:
  - category: Advisory
    durable: true
    lastTransitionTime: "2020-10-29T21:53:45Z"
    message: The migration has completed successfully.
    reason: Final
    status: "True"
    type: Succeeded
  itinerary: Final
  observedDigest: 9834d071975562d5e2c2eb855bca6950711ded8a0e45af5307fa56cd0f5ba3c7
  phase: Completed
  pipeline:
  - completed: "2020-10-29T21:48:14Z"
    message: Completed
    name: Prepare
    started: "2020-10-29T21:45:11Z"
  - completed: "2020-10-29T21:48:49Z"
    message: Completed
    name: Backup
    started: "2020-10-29T21:48:14Z"
  - completed: "2020-10-29T21:50:22Z"
    message: Completed
    name: StageBackup
    started: "2020-10-29T21:48:49Z"
  - completed: "2020-10-29T21:52:24Z"
    message: Completed
    name: StageRestore
    started: "2020-10-29T21:50:22Z"
  - completed: "2020-10-29T21:53:45Z"
    message: Completed
    name: Restore
    started: "2020-10-29T21:52:24Z"
  - completed: "2020-10-29T21:53:45Z"
    message: Completed
    name: Final
    started: "2020-10-29T21:53:45Z"
  startTimestamp: "2020-10-29T21:45:11Z"
```

## QE Test Results 

Commit [#ca28ee8](https://github.com/konveyor/mig-controller/pull/706/commits/ca28ee85c9fdca1a54fe56566291bffe25c93ac7)

Image: quay.io/pranavgaikwad/mig-controller:latest

```
EXECUTION RESULT: OK (smoke/ocp-29946-hooksfailure.yml)
EXECUTION RESULT: OK (smoke/ocp-25986-maxpods.yml)
EXECUTION RESULT: OK (smoke/ocp-29918-hooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24871-datagrid.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24769-cakephp.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34867-imagestreamsstage.yml)
EXECUTION RESULT: OK (smoke/ocp-24659-mysql.yml)
EXECUTION RESULT: OK (smoke/ocp-30240-datavalidation.yml)
EXECUTION RESULT: OK (smoke/ocp-25021-cronjob.yml)
EXECUTION RESULT: OK (smoke/ocp-27335-velero-version.yml)
EXECUTION RESULT: OK (smoke/ocp-32077-empty-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-24995-role.yml)
EXECUTION RESULT: OK (smoke/ocp-31358-restartrestic-newer-than-310.yml)
EXECUTION RESULT: OK (smoke/ocp-26160-max-pvs.yml)
EXECUTION RESULT: OK (smoke/ocp-24686-project.yml)
EXECUTION RESULT: OK (smoke/ocp-25022-cronjob-quiesced.yml)
EXECUTION RESULT: OK (smoke/ocp-24706-basicvolmig.yml)
EXECUTION RESULT: OK (smoke/ocp-24797-mongodb.yml)
EXECUTION RESULT: OK (smoke/ocp-29073-internalis.yml)
EXECUTION RESULT: OK (smoke/ocp-34821-increase-data.yml)
EXECUTION RESULT: OK (smoke/ocp-29947-customhooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24787-redis.yml)
EXECUTION RESULT: OK (smoke/ocp-28967-migplan-naming.yml)
EXECUTION RESULT: OK (smoke/ocp-24730-django.yml)
EXECUTION RESULT: OK (smoke/ocp-34824-decrease-data.yml)
EXECUTION RESULT: OK (smoke/ocp-31309-quotanoattach.yml)
EXECUTION RESULT: OK (smoke/ocp-34703-crd.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24997-configmap.yml)
EXECUTION RESULT: OK (smoke/ocp-33378-max-pvs-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-26032-maxns.yml)
EXECUTION RESULT: OK (smoke/ocp-25090-jobs.yml)
EXECUTION RESULT: OK (smoke/ocp-32189-cronjob-with-pvc.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28021-istags.yml)
EXECUTION RESULT: OK (smoke/ocp-29871-resquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33393-bz1845569.yml)
EXECUTION RESULT: OK (smoke/ocp-33423-hooksnosa.yml)
EXECUTION RESULT: OK (smoke/ocp-32834-pvc-terminating.yml)
EXECUTION RESULT: OK (smoke/ocp-25000-sets.yml)
EXECUTION RESULT: OK (smoke/ocp-31339-pvc-not-connected.yml)
EXECUTION RESULT: OK (smoke/ocp-25212-initcont.yml)
EXECUTION RESULT: OK (smoke/ocp-33375-maxpods-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33379-maxns-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-24872-mssql.yml)
```

### CODE REVIEW

File: pkg/controller/migmigration/migrate.go [hunk 1]
```diff
@@ -92,15 +91,12 @@ func (r *ReconcileMigMigration) migrate(migration *migapi.MigMigration) (time.Du
...
-	step, n, total := task.Itinerary.progressReport(task.Phase)
-	message := fmt.Sprintf(RunningMessage, n, total)
 	migration.Status.SetCondition(migapi.Condition{
 		Type:     Running,
 		Status:   True,
-		Reason:   step,
+		Reason:   task.Step,
 		Category: Advisory,
-		Message:  message,
+		Message:  PhaseDescriptions[task.Phase],
```
Comment (djwhatle) [inline pkg/controller/migmigration/migrate.go:99]: We need to verify that a PhaseDescription for the Phase exists before setting it here.

File: pkg/controller/migmigration/task.go [hunk 2]
```diff
@@ -89,139 +99,201 @@ type Itinerary struct {
...
-		{phase: Completed},
+		{
+			Name: StepFinal,
+			Phases: []Phase{
+				{Name: Canceling},
+				{Name: DeleteBackups},
+				{Name: DeleteRestores},
+				{Name: EnsureStagePodsDeleted, all: HasStagePods},
+				{Name: EnsureAnnotationsDeleted, any: HasPVs | HasISs},
+				{Name: DeleteMigrated},
```
Comment (djwhatle) [inline pkg/controller/migmigration/task.go:249]: @pranavgaikwad we've recently trimmed a few steps from the `CancelItinerary` with our rollback on demand changes. See the diff, we should match the phase list from master.

File: pkg/controller/migmigration/task.go [hunk 3]
```diff
@@ -89,139 +99,201 @@ type Itinerary struct {
...
-		{phase: Completed},
+		{
+			Name: StepFinal,
+			Phases: []Phase{
+				{Name: MigrationFailed},
+				{Name: EnsureStagePodsDeleted, all: HasStagePods},
+				{Name: EnsureAnnotationsDeleted, any: HasPVs | HasISs},
+				{Name: DeleteMigrated},
+				{Name: EnsureMigratedDeleted},
+				{Name: UnQuiesceApplications, all: Quiesce},
```
Comment (djwhatle) [inline pkg/controller/migmigration/task.go:270]: Same here, steps were trimmed from failed.

File: pkg/controller/migmigration/task.go (hunk 2 above)
Comment (pranavgaikwad) [reply pkg/controller/migmigration/task.go:249]: Thank you for pointing that. I missed this in a rebase

**Comment by pranavgaikwad** [comment]
QE Test Results:

```
EXECUTION RESULT: OK (smoke/ocp-29946-hooksfailure.yml)
EXECUTION RESULT: OK (smoke/ocp-25986-maxpods.yml)
EXECUTION RESULT: OK (smoke/ocp-29918-hooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24871-datagrid.yml)
EXECUTION RESULT: OK (smoke/ocp-24769-cakephp.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34867-imagestreamsstage.yml)
EXECUTION RESULT: OK (smoke/ocp-24659-mysql.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-30240-datavalidation.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-25021-cronjob.yml)
EXECUTION RESULT: OK (smoke/ocp-27335-velero-version.yml)
EXECUTION RESULT: OK (smoke/ocp-32077-empty-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-24995-role.yml)
EXECUTION RESULT: OK (smoke/ocp-31358-restartrestic-newer-than-310.yml)
EXECUTION RESULT: OK (smoke/ocp-26160-max-pvs.yml)
EXECUTION RESULT: OK (smoke/ocp-24686-project.yml)
EXECUTION RESULT: OK (smoke/ocp-25022-cronjob-quiesced.yml)
EXECUTION RESULT: OK (smoke/ocp-24706-basicvolmig.yml)
EXECUTION RESULT: OK (smoke/ocp-24797-mongodb.yml)
EXECUTION RESULT: OK (smoke/ocp-29073-internalis.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-34821-increase-data.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-29947-customhooks.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24787-redis.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28967-migplan-naming.yml)
EXECUTION RESULT: OK (smoke/ocp-24730-django.yml)
EXECUTION RESULT: OK (smoke/ocp-34824-decrease-data.yml)
EXECUTION RESULT: OK (smoke/ocp-31309-quotanoattach.yml)
EXECUTION RESULT: OK (smoke/ocp-34703-crd.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24997-configmap.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33378-max-pvs-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-26032-maxns.yml)
EXECUTION RESULT: OK (smoke/ocp-25090-jobs.yml)
EXECUTION RESULT: OK (smoke/ocp-32189-cronjob-with-pvc.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-28021-istags.yml)
EXECUTION RESULT: OK (smoke/ocp-29871-resquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33393-bz1845569.yml)
EXECUTION RESULT: OK (smoke/ocp-33423-hooksnosa.yml)
EXECUTION RESULT: OK (smoke/ocp-32834-pvc-terminating.yml)
EXECUTION RESULT: OK (smoke/ocp-25000-sets.yml)
EXECUTION RESULT: OK (smoke/ocp-31339-pvc-not-connected.yml)
EXECUTION RESULT: OK (smoke/ocp-25212-initcont.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33375-maxpods-equalquota.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-33379-maxns-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-24872-mssql.yml)
```

Some of the above failures were due to api server timeouts, others were just due to migration taking longer time than the time configured for retries. I ran failed tests again one by one, all but one succeded when tried one at a time. `ocp-28021-istags.yml` failed in both cases.

**Comment by alaypatel07** [comment]
> ocp-28021-istags.yml failed in both cases.

This test case is expected to fail. There is an open bug for this.

File: pkg/controller/migmigration/description.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"PreBackupHooksFailed":            "Migration failed while running user-defined pre-backup hooks.",
+	"PostBackupHooksFailed":           "Migration failed while running user-defined post-backup hooks.",
+	"PreRestoreHooksFailed":           "Migration failed while running user-defined pre-restore hooks.",
+	"PostRestoreHooksFailed":          "Migration failed while running user-defined post-restore hooks.",
+	"EnsureInitialBackup":             "Creating initial Velero backup.",
+	"InitialBackupCreated":            "Waiting for initial Velero backup to complete.",
+	"InitialBackupFailed":             "Migration failed during initial Velero backup.",
+	"AnnotateResources":               "Adding migration annotations and labels to PVs, PVCs, Pods, ImageStreams, and Namespaces. Annotations and labels provide migration instructions to Velero, Velero Plugins and Restic.",
+	"EnsureStagePodsFromRunning":      "Creating Stage Pods and mounting PVC data from running Pods.",
+	"EnsureStagePodsFromTemplates":    "Creating Stage Pods and mounting PVC data from DeploymentTemplates, DeploymentConfigs, ReplicationControllers, DaemonSets, StatefulSets, ReplicaSets, CronJobs and Jobs.",
```
Comment (alaypatel07) [inline pkg/controller/migmigration/description.go:21]: Suggest rewording this to: "Creating Stage Pods and mounting PVC using pod templates from Deployments, DeploymentConfigs, ReplicationControllers, DaemonSets, StatefulSets, ReplicaSets, CronJobs and Jobs."

Comment (djwhatle) [reply pkg/controller/migmigration/description.go:21]: 👍  on the change, just need to capitalize `Pod` per UX/docs team guidance on these descriptions.

**Comment by pranavgaikwad** [comment]
@djwhatle I added old `Phase: x/y` progress report back to the `message` field along with descriptions. The goal is to not break UI until new changes come in. I added a TODO comment and will create a new issue to remove those when UI changes take place.

**Comment by djwhatle** [comment]
@pranavgaikwad I was going to recommend considering that to enable earlier merge, 👍

**Comment by pranavgaikwad** [comment]
UI seems to be working fine with the current message field:

```
  conditions:
  - category: Advisory
    lastTransitionTime: "2020-10-26T21:00:46Z"
    message: 'Phase: 3/34, Starting refresh on MigPlan, MigStorage and MigCluster
      resources'
    reason: Prepare
    status: "True"
    type: Running
  - category: Required
    lastTransitionTime: "2020-10-26T21:00:31Z"
    message: The migration is ready.
    status: "True"
    type: Ready
```

Screenshot from 2020-10-26 17-01-00

File: pkg/apis/migration/v1alpha1/timed.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+func (r *Timed) Running() bool {
+	return r.MarkedStarted() && !r.MarkedCompleted()
+}
+
+// MarkedCompleted check if completed.
+func (r *Timed) MarkedCompleted() bool {
+	return r.Completed != nil
+}
+
+func (r *Timed) now() *meta.Time {
```
Comment (alaypatel07) [inline pkg/apis/migration/v1alpha1/timed.go:50]: is this function really needed? Can we not use meta.Now() where ever this function is intended to use?

File: pkg/apis/migration/v1alpha1/timed.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+// Timed records started and completed timestamps.
+type Timed struct {
+	// Started timestamp.
+	Started *meta.Time `json:"started,omitempty"`
+	// Completed timestamp.
+	Completed *meta.Time `json:"completed,omitempty"`
+}
+
+// MarkReset mark as reset.
+func (r *Timed) MarkReset() {
```
Comment (alaypatel07) [inline pkg/apis/migration/v1alpha1/timed.go:14]: is this function really needed? will creating a new `Timed{}` object instead achieve the same result?

File: pkg/apis/migration/v1alpha1/timed.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	r.Started = nil
+	r.Completed = nil
+}
+
+// MarkStarted mark as started.
+func (r *Timed) MarkStarted() {
+	if r.Started == nil {
+		r.Started = r.now()
+		r.Completed = nil
```
Comment (alaypatel07) [inline pkg/apis/migration/v1alpha1/timed.go:23]: pull this out of the if condition? Completed should be reset `nil` every time this function is called right?

File: pkg/controller/migmigration/task.go [hunk 8]
```diff
@@ -44,6 +44,7 @@ const (
 	EnsureStagePodsFromTemplates    = "EnsureStagePodsFromTemplates"
 	EnsureStagePodsFromOrphanedPVCs = "EnsureStagePodsFromOrphanedPVCs"
 	StagePodsCreated                = "StagePodsCreated"
+	StagePodsFailed                 = "StagePodsFailed"
```
Comment (djwhatle) [inline pkg/controller/migmigration/task.go:47]: I believe this was removed from master

**Review by djwhatle — APPROVED** [review]
LGTM! Pipeline view looks good.

File: pkg/apis/migration/v1alpha1/timed.go (hunk 6 above)
Comment (pranavgaikwad) [reply pkg/apis/migration/v1alpha1/timed.go:14]: @alaypatel07 we need this timed.go file as-is for future pipeline improvements. This is taken from virt-controller

**Review by alaypatel07 — COMMENTED** [review]
Added some initial comments, will continue the review in some time. In general, I have a design question. I could envision an alternative design of the same end goal which would not involve a lot of the Itinerary switch phase switching logic:

1. Introduce a new CR, MigMigrationPipeline. This CR is intended to be used by the UI only
2. The controller for this new CR will watch on MigMigration and MigMigrationPipeline objects.
3. Any time MigMigration CR changes its phase from in its Itinerary, this new controller will have a mapping of Itinerary phase to Step. The MigMigrationPipeline CR will update its status to reflect the right step for the phase.
4. The UI will only query the MigMigrationPipeline for pipeline view and MigMigration CR for a detailed view.

Questions:
1. Was this considered and rejected?
2. Do you say any major problem with the alternative?

I will say this, I haven't spent as much time thinking about this as Pranav has so I might be missing a lot of implementation detail. Also, know that this is blocking UI for 1.4.0. But just some thoughts to explore

File: pkg/apis/migration/v1alpha1/timed.go (hunk 7 above)
Comment (pranavgaikwad) [reply pkg/apis/migration/v1alpha1/timed.go:23]: not really, MarkStarted() should only mark a Step started, it shouldn't worry about whether a step is completed or not. 
but when it does start a step, it should make sure the completed is not set by mistake

**Comment by djwhatle** [comment]
@alaypatel07 I don't like the idea of creating a degree of separation between the migmigration object and the user being able to see progress.

**Comment by alaypatel07** [comment]
@djwhatle that is why we have the UI for it :) if you are really at the CLI for debugging, you would really be looking at the phase level detail rather than the pipeline view right?

**Comment by djwhatle** [comment]
> you would really be looking at the phase level detail and ignoring the MigMigrationPipeline right? The step is just a syntax wrapping to make the learning experience smother in UI

No, this is not true. The pipeline view is where progress on Backups, Restores, PVBs, PVRs will be reported.

File: pkg/apis/migration/v1alpha1/timed.go (hunk 7 above)
Comment (alaypatel07) [reply pkg/apis/migration/v1alpha1/timed.go:23]: aah I see. So you want to make MarkStarted idempotent, cool

**Comment by alaypatel07** [comment]
> No, this is not true. The pipeline view is where progress on Backups, Restores, PVBs, PVRs will be reported.

In the alternate design that could still be reported at phase level granularity and carried forward to the pipeline for the UI. So a user will still be able to see the progress of the phase that includes velero resource. Anyway, too late for the design discussion I guess, I would be better off focussing on making this one ready for merge

File: pkg/apis/migration/v1alpha1/timed.go (hunk 5 above)
Comment (alaypatel07) [reply pkg/apis/migration/v1alpha1/timed.go:50]: since we need this as it, marking the comment resolved

**Review by alaypatel07 — APPROVED** [review]
VISIACK.

Didnt get a chance to try this but assuming others have.

**Review by djwhatle — APPROVED** [review]
LGTM. Just tested this. It works well and doesn't break mig-ui while we transition over to reading data out of pipeline. Nice work integrating this @pranavgaikwad
